### PR TITLE
Added devtools support to fetch for XHR

### LINKED ISSUE #11774: Integrate developer tools with fetch algorithm
The code in `components/net/fetch/methods.rs` does not report any information about the requests and responses to the developer tools. Since XMLHttpRequest now relies on the fetch code instead of `components/net/http_loader.rs`, this means that they no longer appear in the developer tools output. The code in `methods.rs` should accept an optional sender to communicate with the devtools server like `http_loader.rs` does, and replicate the notifications that that code sends.

Code: `components/net/fetch/methods.rs`, `components/net/http_loader.rs`, `components/net/resource_thread.rs`

### PR DESCRIPTION
**Pull request by avadacatavra**
Added devtools support for fetch for XHR, but devtools can't show the request in the XHR tab. I've attached a picture of an XHR request in devtools.

screenshot 2016-07-19 11 07 55

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11774  (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because I don't know how to automate devtools testing for this

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12502)

### CODE REVIEW

File: components/net/fetch/methods.rs [hunk 1]
```diff
@@ -52,14 +54,18 @@ pub struct FetchContext {
...
+             devtools_chan: Option<Sender<DevtoolsControlMsg>>) -> Response {
+    fetch_with_cors_cache(request, &mut CORSCache::new(), target, context, devtools_chan)
 }
 
 pub fn fetch_with_cors_cache(request: Rc<Request>,
                              cache: &mut CORSCache,
                              target: &mut Target,
-                             context: FetchContext) -> Response {
+                             context: FetchContext,
+                             devtools_chan: Option<Sender<DevtoolsControlMsg>>) -> Response {
```
Comment (Manishearth) [inline components/net/fetch/methods.rs]: I feel like the devtools_chan can be rolled into the FetchContext

**Comment by avadacatavra** [comment]
CC @jdm

File: components/net/http_loader.rs [hunk 2]
```diff
@@ -601,6 +601,7 @@ fn prepare_devtools_request(request_id: String,
                             now: Tm,
                             connect_time: u64,
                             send_time: u64) -> ChromeToDevtoolsControlMsg {
+    debug!("preparing!!!");
```
Comment (Manishearth) [inline components/net/http_loader.rs]: might want to clean these up 😄 

It's fine to leave debug statements around, but try to have them print something useful, e.g. `prepare_devtools_request: <url>` or something

Comment (Manishearth) [inline components/net/http_loader.rs]: (This applies to the other debug statements in this PR, too)

File: components/net/resource_thread.rs [hunk 3]
```diff
@@ -586,7 +587,7 @@ impl CoreResourceManager {
             // todo service worker stuff
             let mut target = Some(Box::new(sender) as Box<FetchTaskTarget + Send + 'static>);
             let context = FetchContext { state: http_state, user_agent: ua };
-            fetch(Rc::new(request), &mut target, context);
+            fetch(Rc::new(request), &mut target, context, dc);            //FIXME put a devtools channel here
```
Comment (Manishearth) [inline components/net/resource_thread.rs]: fixme was fixed?

**Comment by Manishearth** [comment]
r? @jdm 

Overall lgtm, with the few nits I mentioned here and in IRC. I think I've got a fix for the XHR thing ready.

**Comment by Manishearth** [comment]
https://github.com/Manishearth/servo/commit/a388da3061601cbcf557de9f49d06a919ac218b3 should work for getting XHR to work. However, I'm unable to test it, since I can't get any network requests to show in my devtools pane.

**Comment by jdm** [comment]
Looking good! The unit tests in tests/unit/net/fetch.rs will need to be updated too - `./mach test-unit -p net`.
-S-awaiting-review +S-tests-failed +S-needs-code-changes 

---

Reviewed 5 of 5 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 961 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12502#-KN4BqDHxEMamNMZI8Iu:-KN4BqDHxEMamNMZI8Iv:204571324) ([raw file](https://github.com/servo/servo/blob/a70e807dd19455188f7346ab49e1c5dc62b7995e/components/net/fetch/methods.rs#L961)):*
> ```Rust
>     let cancellation_listener = CancellationListener::new(None);
> 
>     let request_id = "";
> ```

We should generate an ID like https://dxr.mozilla.org/servo/rev/5f702d507c908348382d376f43e3cfc08094b517/components/net/http_loader.rs#968

---

*[components/net/fetch/methods.rs, line 984 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12502#-KN4Dp_m41f6FhadKlNJ:-KN4Dp_m41f6FhadKlNK:-1454518423) ([raw file](https://github.com/servo/servo/blob/a70e807dd19455188f7346ab49e1c5dc62b7995e/components/net/fetch/methods.rs#L984)):*
> ```Rust
>             let done_sender = done_chan.as_ref().map(|ch| ch.0.clone());
>             let devtools_sender = devtools_chan.clone();
>             let metadata = meta.clone();
> ```

Let's clone the fields we actually need, instead. Then we don't need to clone them later.

---

*[components/net/fetch/methods.rs, line 1022 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12502#-KN4C6GOH9D6dAiCnmSZ:-KN4C6GP0NNakVuDJvfc:-1741494907) ([raw file](https://github.com/servo/servo/blob/a70e807dd19455188f7346ab49e1c5dc62b7995e/components/net/fetch/methods.rs#L1022)):*
> ```Rust
>                         }
>                         if let Some(m) = msg {
>                             send_request_to_devtools(m, devtools_sender.as_ref().unwrap());
> ```

Let's do this before the loop.

---

*[components/net/fetch/methods.rs, line 1032 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12502#-KN4Cd2gw3YbPUCfVJQU:-KN4Cd2gw3YbPUCfVJQV:-1517735050) ([raw file](https://github.com/servo/servo/blob/a70e807dd19455188f7346ab49e1c5dc62b7995e/components/net/fetch/methods.rs#L1032)):*
> ```Rust
>                             devtools_sender, request_id.into(),
>                             metadata.headers.clone(), metadata.status.clone(),
>                             pipeline_id);
> ```

nit: the arguments should be indented one level.

---

*[components/net/fetch/methods.rs, line 1052 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12502#-KN4Du7y6L0tUSPw-klE:-KN4Du7y6L0tUSPw-klF:-2072554518) ([raw file](https://github.com/servo/servo/blob/a70e807dd19455188f7346ab49e1c5dc62b7995e/components/net/fetch/methods.rs#L1052)):*
> ```Rust
>         }
>     };
> ```

nit: remove this extra newline.

---

*[components/net_traits/request.rs, line 288 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12502#-KN4E7Ydc8nm0K_32Ysd:-KN4E7Ydc8nm0K_32Yse:-1706084616) ([raw file](https://github.com/servo/servo/blob/a70e807dd19455188f7346ab49e1c5dc62b7995e/components/net_traits/request.rs#L288)):*
> ```Rust
>             redirect_count: Cell::new(0),
>             response_tainting: Cell::new(ResponseTainting::Basic),
>             pipeline_id: Cell::new(Some(PipelineId::new())),
> ```

We should pass the pipeline id in as an argument, instead.

---

*[components/script/dom/xmlhttprequest.rs, line 609 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12502#-KN4EMZX4a5cWOlC76Nx:-KN4EMZYsmCQSd8AHl80:1090565188) ([raw file](https://github.com/servo/servo/blob/a70e807dd19455188f7346ab49e1c5dc62b7995e/components/script/dom/xmlhttprequest.rs#L609)):*
> ```Rust
>             referer_url: self.referrer_url.clone(),
>             referrer_policy: self.referrer_policy.clone(),
>             pipeline_id: self.pipeline_id().clone(),
> ```

This clone shouldn't be necessary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12502#-:-KN4FTXQ6pQZQ9J6ramp:-376932876)*

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes +S-fails-travis 

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[components/net/fetch/methods.rs, line 1095 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12502#-KNJT0K9WiIN-FCLupn3:-KNJT0KAKDqyCpHNDwp9:1731282260) ([raw file](https://github.com/servo/servo/blob/f6572fe55022b8920e41abbdc97e83398b8e56cd/components/net/fetch/methods.rs#L1095)):*
> ```Rust
>     // Step 1
>     let mut preflight = Request::new(request.current_url(), Some(request.origin.borrow().clone()), 
>                                      false, Some(PipelineId::new()));
> ```

The pipeline id should be passed in as an argument instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12502#-:-KNJT8VqYENIpfIfwDbu:1516864919)*

**Comment by avadacatavra** [comment]
@jdm I was confused on what you meant by "Let's clone the fields we actually need, instead. Then we don't need to clone them later."

will be adding the unit tests today/tomorrow, and I put @Manishearth's fix for identifying XHR in

**Comment by jdm** [comment]
We should do `let status = metadata.status.clone()` instead of cloning the entire metadata value.

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12502#-:-KNXJAoNIeasmhSUjhFA:1458266033)*

**Comment by avadacatavra** [comment]
@jdm I added a unit test for devtools fetch

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---

*[tests/unit/net/fetch.rs, line 776 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12502#-KNbRg4YY52-QnUxLlxz:-KNbRg4YY52-QnUxLly-:781904383) ([raw file](https://github.com/servo/servo/blob/c137e97b9517431e6c562008b5c80fa8c7c2df45/tests/unit/net/fetch.rs#L776)):*
> ```Rust
> 
>     let request = Request::new(url.clone(), Some(origin), false, Some(pipeline_id));
>     let _ = fetch_sync(request, Some(devtools_chan));
> ```

I bet this is failing to get a valid response because it's making an actual request to https://mozilla.com and our tests don't support HTTPS, so `expect_devtools_http_request` waits forever and the test run times out. We'll need to use `make_server` like the other tests in this file.

---

*[tests/unit/net/http_loader.rs, line 40 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12502#-KNbSHCNQ5OGijlIb2aj:-KNbSHCNQ5OGijlIb2ak:-1623723411) ([raw file](https://github.com/servo/servo/blob/c137e97b9517431e6c562008b5c80fa8c7c2df45/tests/unit/net/http_loader.rs#L40)):*
> ```Rust
> const DEFAULT_USER_AGENT: &'static str = "Test-agent";
> 
> pub struct HttpTest;
> ```

Why are all of these types and methods being made public? Many don't seem to be used by the fetch tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12502#-:-KNbSWJDww3KXqqcNDrW:1458266033)*

**Comment by jdm** [comment]
Great! There are still a couple unresolved review comments, though.
-S-awaiting-review -S-fails-travis +S-needs-code-changes 

---

Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[tests/unit/net/fetch.rs, line 781 \[r6\]](https://reviewable.io:443/reviews/servo/servo/12502#-KNhVCVeJIhP1Wsb_oSz:-KNhVCVfkKfut9zIxMu1:738962002) ([raw file](https://github.com/servo/servo/blob/42d1fcebff3ca695f0a8efe58d117dc07109c561/tests/unit/net/fetch.rs#L781)):*
> ```Rust
>     let mut request = Request::new(url.clone(), Some(origin), false, None);
>     let pipeline_id = PipelineId::fake_root_pipeline_id();
>     request.pipeline_id = Cell::new(Some(pipeline_id));
> ```

We can pass `Some(pipeline_id)` as the last argument of Request::new, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12502#-:-KNhVQ4DpvQEcQwUHvIC:-852144824)*

**Comment by avadacatavra** [comment]
@jdm sorry about the Request arguments derp :)

**Comment by avadacatavra** [comment]
Looking for a more interesting screenshot :)

screenshot 2016-07-28 11 25 36

**Comment by avadacatavra** [comment]
screenshot 2016-07-28 12 00 25

**Comment by jdm** [comment]
-S-needs-rebase -S-awaiting-review +S-needs-code-changes +S-fails-travis 

---

Reviewed 4 of 4 files at r7.
Review status: 8 of 9 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[components/net/fetch/methods.rs, line 1095 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12502#-KNJT0K9WiIN-FCLupn3:-KNlKRitgfIrdI1q0TZF:1205187691) ([raw file](https://github.com/servo/servo/blob/f6572fe55022b8920e41abbdc97e83398b8e56cd/components/net/fetch/methods.rs#L1095)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> The pipeline id should be passed in as an argument instead.
</details>
This still needs to be addressed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12502#-:-KNlK_z7dPifVd5lbVvm:1517581586)*

**Comment by jdm** [comment]
Almost there! Also, there are a couple remaining debug! uses in http_loader.rs that were added in this PR; let's remove them.
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 1 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/net/http_loader.rs, line 632 \[r8\]](https://reviewable.io:443/reviews/servo/servo/12502#-KNlT5Fshp1vtCxIR5MV:-KNlT5Fshp1vtCxIR5MW:-1182753582) ([raw file](https://github.com/servo/servo/blob/3ef2709a9116e937a38429474931acba9c3e7a75/components/net/http_loader.rs#L632)):*
> ```Rust
>                              status: Option<RawStatus>,
>                              pipeline_id: PipelineId) {
>     if devtools_chan.is_none() {
> ```

I'd prefer to revert this change; the previous interface that accepted a channel worked fine.

---

*[components/net/http_loader.rs, line 1099 \[r8\]](https://reviewable.io:443/reviews/servo/servo/12502#-KNlTElFIjaghsDvBEXC:-KNlTElGTTCydmu1Lv8-:1384968219) ([raw file](https://github.com/servo/servo/blob/3ef2709a9116e937a38429474931acba9c3e7a75/components/net/http_loader.rs#L1099)):*
> ```Rust
>         // TODO: Send this message even when the load fails?
>         if let Some(pipeline_id) = load_data.pipeline_id {
>            // if let Some(ref chan) = devtools_chan {
> ```

Revert this, please.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12502#-:-KNlTPmoapWxTBoEaLx2:1935589987)*

**Comment by avadacatavra** [comment]
WTF I know I removed those!

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[tests/unit/net/http_loader.rs, line 40 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12502#-KNbSHCNQ5OGijlIb2aj:-KNlUkLkG4AzQGIBuTFE:-499146009) ([raw file](https://github.com/servo/servo/blob/c137e97b9517431e6c562008b5c80fa8c7c2df45/tests/unit/net/http_loader.rs#L40)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> Why are all of these types and methods being made public? Many don't seem to be used by the fetch tests.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12502#-:-KNlUrFTw7uYjpu82EQl:586311757)*

**Comment by jdm** [comment]
-S-awaiting-review -S-fails-travis +S-needs-code-changes 

---

Reviewed 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[components/net/http_loader.rs, line 627 \[r10\]](https://reviewable.io:443/reviews/servo/servo/12502#-KNlhoRIWvNZTQ9Idul4:-KNlhoRJaaddJeCVfG65:5172382) ([raw file](https://github.com/servo/servo/blob/17da4ac9240e3499155e888e36c8bf73dd706bb8/components/net/http_loader.rs#L627)):*
> ```Rust
> }
> 
> pub fn send_response_to_devtools(devtools_chan: Sender<DevtoolsControlMsg>,
> ```

&?

---

*[components/net/http_loader.rs, line 1094 \[r10\]](https://reviewable.io:443/reviews/servo/servo/12502#-KNlhr7uHD2FG7rLkUSS:-KNlhr7uHD2FG7rLkUST:1090565188) ([raw file](https://github.com/servo/servo/blob/17da4ac9240e3499155e888e36c8bf73dd706bb8/components/net/http_loader.rs#L1094)):*
> ```Rust
>             if let Some(ref chan) = devtools_chan {
>                 send_response_to_devtools(
>                     chan.clone(), request_id,
> ```

This clone shouldn't be necessary.

---

*[components/net/fetch/methods.rs, line 991 \[r10\]](https://reviewable.io:443/reviews/servo/servo/12502#-KNli-Kz95RGebF6_wnk:-KNli-Kz95RGebF6_wnl:1090565188) ([raw file](https://github.com/servo/servo/blob/17da4ac9240e3499155e888e36c8bf73dd706bb8/components/net/fetch/methods.rs#L991)):*
> ```Rust
>                         if let Some(ref sender) = devtools_sender {
>                             if let Some(m) = msg {
>                                 send_request_to_devtools(m, &sender.clone());
> ```

This clone shouldn't be necessary.

---

*[components/net/fetch/methods.rs, line 998 \[r10\]](https://reviewable.io:443/reviews/servo/servo/12502#-KNlhxcaaye7rLgmarzH:-KNlhxcaaye7rLgmarzI:1090565188) ([raw file](https://github.com/servo/servo/blob/17da4ac9240e3499155e888e36c8bf73dd706bb8/components/net/fetch/methods.rs#L998)):*
> ```Rust
>                             if let Some(pipeline_id) = pipeline_id {
>                                 send_response_to_devtools(
>                                     sender.clone(), request_id.into(),
> ```

This clone shouldn't be necessary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12502#-:-KNli71gnuZNMsAeNSp8:1550941781)*

**Comment by jdm** [comment]
@bors-servo: r+

---

Reviewed 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12502#-:-KNrDbLCnFTzuqKcKcIq:-317921516)*

**Pull request merged**